src/greynirseq/utils/preprocessing/filters.py

```diff
@@ -0,0 +1,997 @@
+    except TypeError  as e:
+        return True
+    return len(isl_toks) >= DEFAULT_MIN_WORD_COUNT and len(eng_toks) >= DEFAULT_MIN_WORD_COUNT
+
+
+class Gather:
+
+    _store = None
+    _prev_store = None
+    _initialized_with_file = False
+
+    @classmethod
+    def _idemp_init(cls):
+        """ Idempotently initialize, if gather data exists from last run
+            it will be used to filter/transform. Otherwise, pipeline needs
+            to be executed twice. """
+        raise NotImplementedError
+
+    @classmethod
+    def gather(cls, ex):
+        raise NotImplementedError
+
+    @classmethod
+    def save_to_file(cls):
+        # Write gathered data to file
+        # This depends on the data structure the gatherer uses
+        raise NotImplementedError
+
+
+class MinFrequency(Gather):
+
+    auto_pass_len = 3  # for numbers and abbreviations
+    min_freq = 2
+    word_prog = RegexCache.compile_rx(r"\b\w+\b")
+    num_prog = RegexCache.compile_rx(r"^\d+$")
+
+    @classmethod
+    def _idemp_init(cls):
+        if cls._store is not None:
+            return
+        try:
+            os.makedirs(_TMP_DIR, exist_ok=True)
+            path = os.path.join(_TMP_DIR, cls.__name__)
+            with open(path, "r") as fh:
+                cls._prev_store = json.load(fh)
+        except (FileNotFoundError, json.decoder.JSONDecodeError) as e:
+            pass
+        cls._store = {"eng": collections.Counter(), "ice": collections.Counter()}
+
+    @classmethod
+    def save_to_file(cls):
+        try:
+            os.makedirs(_TMP_DIR, exist_ok=True)
+            path = os.path.join(_TMP_DIR, cls.__name__)
+            with open(path, "w") as fh:
+                json.dump(cls._store, fh, indent=2)
+        except FileNotFoundError:
+            pass
+
+    @classmethod
+    def gather(cls, ex):
+        cls._idemp_init()
+        ice, eng = ex["is"], ex["en"]
+        ice_words = cls.word_prog.findall(ice)
+        eng_words = cls.word_prog.findall(eng)
+        ice_words = [
+            w.lower()
+            for w in ice_words
+            if len(w) > cls.auto_pass_len and not cls.num_prog.match(w)
+        ]
+        eng_words = [
+            w.lower()
+            for w in eng_words
+            if len(w) > cls.auto_pass_len and not cls.num_prog.match(w)
+        ]
+        cls._store["ice"].update(ice_words)
+        cls._store["eng"].update(eng_words)
+
+        if cls._prev_store:
+            pice = all(
+                cls._prev_store["ice"].get(w, 0) >= cls.min_freq for w in ice_words
+            )
+            peng = all(
+                cls._prev_store["eng"].get(w, 0) >= cls.min_freq for w in eng_words
+            )
+            return pice and peng
+        return True
+
+
+
+@register_filter
+def wrong_quotes(ex):
+    ice, eng = ex["is"], ex["en"]
+
+    QUOTES = set(QUOTE_LIKE)
+    QUOTES.remove("'")
+    QUOTES.remove('"')
+    QUOTES.remove(",")
+    for q in ICE_QUOTE.ALL:
```

Comment:
This can be done in a more elegant way in Python, but never mind

Reply:
Changed